### PR TITLE
Implement shmup converter

### PR DESCRIPTION
**Pull request by elhmn**
This pull request adds an initial shmup converter.

- [x] Implements a `sha256` hash of the file path, that is used to map git data with files.
- [x] Convert data extracted from the git and code structure into scenes and entities.

The data converted are stored under `./tmp/scanner/repo-slug/shmup-converted.json`

**Sample of extracted data for https://github.com/elhmn/qautomata**
<details>
<summary> shmup-converted.json data</summary>

```json
{
  "15d4b25484d5489424549747d94dacf9c38314a1": {
    "entities": {
      "8ace2c75c3f4f53b0d64b8887bcd916490da8443": {
        "id": "8ace2c75c3f4f53b0d64b8887bcd916490da8443",
        "scene_id": "8ace2c75c3f4f53b0d64b8887bcd916490da8443",
        "name": "README.md",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "088ba6ba7d345b76aa2b8dc021dd25e1323189b3": {
        "id": "088ba6ba7d345b76aa2b8dc021dd25e1323189b3",
        "scene_id": "088ba6ba7d345b76aa2b8dc021dd25e1323189b3",
        "name": ".gitignore",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "b00c5cd2b4f4fe957d03a6ad5c02daaff4cca174": {
        "id": "b00c5cd2b4f4fe957d03a6ad5c02daaff4cca174",
        "scene_id": "b00c5cd2b4f4fe957d03a6ad5c02daaff4cca174",
        "name": "Cargo.toml",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "f288702d2fa16d3cdf0035b15a9fcbc552cd88e7": {
        "id": "f288702d2fa16d3cdf0035b15a9fcbc552cd88e7",
        "scene_id": "f288702d2fa16d3cdf0035b15a9fcbc552cd88e7",
        "name": "LICENSE",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      }
    },
    "sub_scenes": [
      "2a8d05b4917ac67e68be343f85de822b3e1f5826",
      "a0d67a1bf5035faee0918acbe7d624621208142f",
      "3c23db1a468e82e512378f1e4ed9734ad82c0e51"
    ]
  },
  "a0d67a1bf5035faee0918acbe7d624621208142f": {
    "entities": {
      "74bacd46e074d633fa920dd8ef8f59d8b0868c71": {
        "id": "74bacd46e074d633fa920dd8ef8f59d8b0868c71",
        "scene_id": "74bacd46e074d633fa920dd8ef8f59d8b0868c71",
        "name": "state1.json",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "eb662493e4e0d0d6cbb4b38300f1b128401a4e61": {
        "id": "eb662493e4e0d0d6cbb4b38300f1b128401a4e61",
        "scene_id": "eb662493e4e0d0d6cbb4b38300f1b128401a4e61",
        "name": "state_2_adjacent_cells.json",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "9a2bd39800d4cb34c75515c1caf7a1ada5e17c9d": {
        "id": "9a2bd39800d4cb34c75515c1caf7a1ada5e17c9d",
        "scene_id": "9a2bd39800d4cb34c75515c1caf7a1ada5e17c9d",
        "name": "state_2_diagonal_cells.json",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "1e1fc315c7aaa87f553a787dbfddc4a0e0b36a75": {
        "id": "1e1fc315c7aaa87f553a787dbfddc4a0e0b36a75",
        "scene_id": "1e1fc315c7aaa87f553a787dbfddc4a0e0b36a75",
        "name": "state_grid_20.json",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "3802a58b086884c4e6b7620ac6f4d148c484b333": {
        "id": "3802a58b086884c4e6b7620ac6f4d148c484b333",
        "scene_id": "3802a58b086884c4e6b7620ac6f4d148c484b333",
        "name": "state2.json",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      }
    },
    "sub_scenes": []
  },
  "ac902275fa8a404ad7f6ae26be5355fda347e025": {
    "entities": {
      "e614682b782d0f9ad61e8afb522da1e7c17dbdea": {
        "id": "e614682b782d0f9ad61e8afb522da1e7c17dbdea",
        "scene_id": "e614682b782d0f9ad61e8afb522da1e7c17dbdea",
        "name": "gen.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "062b73166ffe85fad3d64fd24c195a7364f1454b": {
        "id": "062b73166ffe85fad3d64fd24c195a7364f1454b",
        "scene_id": "062b73166ffe85fad3d64fd24c195a7364f1454b",
        "name": "run.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      }
    },
    "sub_scenes": []
  },
  "59181b054e8822507c756175dcfdc55575e6bbd3": {
    "entities": {
      "efcc8a3fbc4c7e147c7969637bb087511216cdd8": {
        "id": "efcc8a3fbc4c7e147c7969637bb087511216cdd8",
        "scene_id": "efcc8a3fbc4c7e147c7969637bb087511216cdd8",
        "name": "grid.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.25,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      }
    },
    "sub_scenes": []
  },
  "2a8d05b4917ac67e68be343f85de822b3e1f5826": {
    "entities": {},
    "sub_scenes": [
      "09a747d69ec87c52a782be97c5610d40941dd9bb"
    ]
  },
  "35a0d5bb7f00b3af2aa166606ef7b3c14d59f989": {
    "entities": {
      "12a1731b33c564311e60306a30a0bb62671aba8f": {
        "id": "12a1731b33c564311e60306a30a0bb62671aba8f",
        "scene_id": "12a1731b33c564311e60306a30a0bb62671aba8f",
        "name": "interference.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "12b7108bb05ddd6d5d87d3e9841cc3a9117b659e": {
        "id": "12b7108bb05ddd6d5d87d3e9841cc3a9117b659e",
        "scene_id": "12b7108bb05ddd6d5d87d3e9841cc3a9117b659e",
        "name": "types.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.5,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "fe1bd960c4d2cad724a90d8b867c5257fd26c1f3": {
        "id": "fe1bd960c4d2cad724a90d8b867c5257fd26c1f3",
        "scene_id": "fe1bd960c4d2cad724a90d8b867c5257fd26c1f3",
        "name": "files.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "0d38104c0f7af25ea67b5601fc8ea7fd611a3f87": {
        "id": "0d38104c0f7af25ea67b5601fc8ea7fd611a3f87",
        "scene_id": "0d38104c0f7af25ea67b5601fc8ea7fd611a3f87",
        "name": "measure.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.5,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "b41faf3f635eae948bac65716d32fb6c935bd354": {
        "id": "b41faf3f635eae948bac65716d32fb6c935bd354",
        "scene_id": "b41faf3f635eae948bac65716d32fb6c935bd354",
        "name": "step.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.125,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      }
    },
    "sub_scenes": []
  },
  "09a747d69ec87c52a782be97c5610d40941dd9bb": {
    "entities": {
      "41c2c797d02483e13aa2b9030faf6dc66ea737f2": {
        "id": "41c2c797d02483e13aa2b9030faf6dc66ea737f2",
        "scene_id": "41c2c797d02483e13aa2b9030faf6dc66ea737f2",
        "name": "build.yaml",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "0f46b276f7aa7670c82b15365a7d9a9ccbe81a1a": {
        "id": "0f46b276f7aa7670c82b15365a7d9a9ccbe81a1a",
        "scene_id": "0f46b276f7aa7670c82b15365a7d9a9ccbe81a1a",
        "name": "lint.yaml",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "ca1aa4cccfaf5e69fafc5b7f25d1795bed29517d": {
        "id": "ca1aa4cccfaf5e69fafc5b7f25d1795bed29517d",
        "scene_id": "ca1aa4cccfaf5e69fafc5b7f25d1795bed29517d",
        "name": "test.yaml",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 0.1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      }
    },
    "sub_scenes": []
  },
  "3c23db1a468e82e512378f1e4ed9734ad82c0e51": {
    "entities": {
      "b804a6206dee1d40e2e26728bfa52ff712da3efd": {
        "id": "b804a6206dee1d40e2e26728bfa52ff712da3efd",
        "scene_id": "b804a6206dee1d40e2e26728bfa52ff712da3efd",
        "name": "main.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "0aae6eb768a0f464f1bec889f0eadf16e7d9b4c4": {
        "id": "0aae6eb768a0f464f1bec889f0eadf16e7d9b4c4",
        "scene_id": "0aae6eb768a0f464f1bec889f0eadf16e7d9b4c4",
        "name": "universe.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": null,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "63b748ee9af19936bebc01409c87a36e445cf61f": {
        "id": "63b748ee9af19936bebc01409c87a36e445cf61f",
        "scene_id": "63b748ee9af19936bebc01409c87a36e445cf61f",
        "name": "cmd.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "8b07682f211d072f854cebbed99ed7a14075d48e": {
        "id": "8b07682f211d072f854cebbed99ed7a14075d48e",
        "scene_id": "8b07682f211d072f854cebbed99ed7a14075d48e",
        "name": "engine.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      },
      "462befd8420152213f0993c0201694f7217f5746": {
        "id": "462befd8420152213f0993c0201694f7217f5746",
        "scene_id": "462befd8420152213f0993c0201694f7217f5746",
        "name": "gui.rs",
        "kind": "kind",
        "color": "a color",
        "weapon": "",
        "movement_pattern": "",
        "speed": 1,
        "hp": 1,
        "size": 0,
        "shield": "",
        "destructible": false
      }
    },
    "sub_scenes": [
      "ac902275fa8a404ad7f6ae26be5355fda347e025",
      "59181b054e8822507c756175dcfdc55575e6bbd3",
      "35a0d5bb7f00b3af2aa166606ef7b3c14d59f989"
    ]
  }
}

```
</details>

Addresses https://github.com/elhmn/waking-git/issues/18

### CODE REVIEW

File: src/cmd/scan.rs [hunk 1]
```diff
@@ -32,45 +34,79 @@ pub fn run(args: &RunArgs, conf: config::Config) {
...
+        "{}/{}-{}",
+        dest_folder,
+        converter.name(),
+        CONVERTER_FILE_NAME_PREFIX
+    );
+    let json_data = serde_json::to_string(&data).unwrap_or_else(|_| "".to_string());
+    match store_json_data(json_data, dest_folder, dest_path.clone()) {
+        Ok(_) => (),
+        Err(err) => {
+            return Err(format!("Error: failed to extract repository data: {}", err));
```
Comment (Git-Math) [inline src/cmd/scan.rs]:
```suggestion
            return Err(format!("Error: failed to convert extracted data: {}", err));
```

**Comment by elhmn** [comment]
The converter does not currently use code file data to set the entity's attributes.
This will be done on subsequent pull requests as we feared the pull request would become too big.

File: src/extractor/git.rs [hunk 2]
```diff
@@ -24,12 +25,28 @@ pub struct Metrics {}
 pub struct Blob {
     pub filemode: i32,
     pub name: String,
+    //path is the path to the file/directiry relative
```
Comment (Git-Math) [inline src/extractor/git.rs]:
```suggestion
    //path is the path to the file/directory relative
```

File: src/extractor/git.rs [hunk 3]
```diff
@@ -24,12 +25,28 @@ pub struct Metrics {}
...
+    //path_sha is the sha 256 of the path
+    //this is used to optimize code data lookup
+    pub path_sha: String,
     pub sha: String,
 }
 
 #[derive(Serialize, Default)]
 pub struct Tree {
     pub name: String,
+    //path is the path to the file/directiry relative
```
Comment (Git-Math) [inline src/extractor/git.rs]:
```suggestion
    //path is the path to the file/directory relative
```

**Review by Git-Math — APPROVED** [review]
Nice

**Pull request merged**